scripts/metalk8s_cli/mixins/log.py

```diff
@@ -0,0 +1,251 @@
+            'level': 'DEBUG',
+            'formatter': 'file',
+            'filters': ['remove_ansi'],
+        },
+    },
+    'root': {
+        'handlers': ['console', 'file'],
+        'level': 'DEBUG',
+    },
+}
+
+
+class LoggingCommandMixin(object):
+    """Provide methods to log messages from a command.
+
+    Always log to a file, and log partial info to the console.
+    A logging command should rely on its `log_active_run` and `log_step`
+    context managers within its `run` method.
+    """
+    def __init__(self, args):
+        super(LoggingCommandMixin, self).__init__(args)
+        self._logger = None
+        self.logfile = os.path.abspath(args.logfile)
+        self.verbose = args.verbose
+        self.prepare_logfile()
+        self.configure_logging()
+
+    def configure_logging(self):
+        config = copy.deepcopy(DEFAULT_CONFIG)
+        config['handlers']['file']['filename'] = self.logfile
+        logging.config.dictConfig(config)
+
+    def prepare_logfile(self):
+        if os.path.exists(self.logfile):
+            return
+
+        dirname = os.path.dirname(self.logfile)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        open(self.logfile, 'a').close()
+
+    @property
+    def logger(self):
+        if self._logger is None:
+            name = '.'.join(self.command_name.split())
+            self._logger = logging.getLogger(name)
+        return self._logger
+
+    @contextlib.contextmanager
+    def log_step(self, step_name):
+        """Log a step execution."""
+        self.print_and_log('> {}...'.format(step_name), keep_line=True)
+        start = time.time()
+        try:
+            # Wrapped step should use `print_and_log` for success, and
+            # raise `CommandError` in case of execution failure
+            yield
+        except CommandError as exc:
+            duration = time.time() - start
+            self.print_and_log('{} [{:.2f}s]'.format(
+                _format_ansi('fail', color='red'), duration
+            ))
+            self.logger.error(self._format_error(exc, step_name))
+            # We re-raise here for the main `run` method to interrupt execution
+            raise
+        else:
+            duration = time.time() - start
+            self.print_and_log('{} [{:.2f}s]'.format(
+                _format_ansi('done', color='green'), duration
+            ))
+
+    @contextlib.contextmanager
+    def log_active_run(self):
+        """Open and close a section in the logfile for a command run."""
+        self._init_run_log()
+        try:
+            # Errors should be caught and logged by the wrapped code, and
+            # re-raised as `CommandError`s
+            yield
+        except CommandError:
+            self.print_and_log('The script will now exit')
+            sys.exit(1)
+        finally:
+            self._stop_run_log()
+
+    def print_and_log(self, message, level=logging.INFO, keep_line=False):
```

Comment:
Hmm it's weird to call this function each time you want to log something.

For me this class should inherit from classic Logging class and dynamicaly change the level of the console handler when verbose mode enable

Reply:
> this class should inherit from classic Logging class

WDYM?

> dynamicaly change the level of the console handler when verbose mode enable

Huh, I did this first, then decided otherwise. The rationale is the following: it's a CLI, so there's a big difference between what goes in `stdout` and `stderr`. IMO (probably wrong), _verbosity_ of the script affects what goes in `stdout`, while log messages go in `stderr` (and we may give knobs to configure this logging for users). This way, if commands are supposed to return parsable output, a consumer can easily redirect `stderr` and avoid any risk when parsing (as you mentioned for the `SaltCommandMixin._get_json_from_minion` :wink:).

Now, I wanted to have the `print`-ed messages in the logfile as well, which is why there's this `print_and_log` method.

> it's weird to call this function each time you want to log something.

I think some occasions may justify using "log-only" messages for `INFO` level and below, but for now all of those were user-oriented. Note that errors are not `print`-ed.

Reply:
> Why not create a class `Metalk8sLogger(logging.getLoggerClass())` ?
Then we could use directly the classic `debug`, `info` ...

Why would we need a custom logger? Isn't the native one sufficient? If you want the `LoggingCommandMixin` to **be** a `Logger` subclass, then I disagree, we should favor composition in that kind of situation.

> if you enable verbose mode then you change the handler for console so that log also printed in the console

See above as to why I don't want to do this ;)

> printing of the beginning of the step + result should be done in `log_step` function directly. Instead of `print_and_log` we should `print` and `log` (maybe something with more detail if needed)

OK, I think that's a good middle-ground. I'll need to filter out `INFO` messages from the `console` handler to avoid duplication, but OK.

Reply:
> Why would we need a custom logger? Isn't the native one sufficient? If you want the `LoggingCommandMixin` to **be** a `Logger` subclass, then I disagree, we should favor composition in that kind of situation.

No really needed but if we want to have some special feature this can help but for the moment maybe not needed.

> > if you enable verbose mode then you change the handler for console so that log also printed in the console
> 
> See above as to why I don't want to do this ;)

Which one ?
You can change the log level of an handler without having a subclass of `Logger`

> OK, I think that's a good middle-ground. I'll need to filter out `INFO` messages from the `console` handler to avoid duplication, but OK.

Why ? We should have no duplication here if you never print in a logging function but only log